README.md
[![Actions Status](https://github.com/p5-UV/Alien-libuv/workflows/linux/badge.svg)](https://github.com/p5-UV/Alien-libuv/actions)
[![Actions Status](https://github.com/p5-UV/Alien-libuv/workflows/macos/badge.svg)](https://github.com/p5-UV/Alien-libuv/actions)
[![Actions Status](https://github.com/p5-UV/Alien-libuv/workflows/windows/badge.svg)](https://github.com/p5-UV/Alien-libuv/actions)

# NAME

Alien::libuv - Interface to the libuv library [http://libuv.org](http://libuv.org)

# SYNOPSIS

In your `Makefile.PL`:

```perl
use strict;
use warnings;

use ExtUtils::MakeMaker;
use Config;
use Alien::Base::Wrapper ();

WriteMakefile(
  Alien::Base::Wrapper->new('Alien::libuv')->mm_args2(
    ...
    CONFIGURE_REQUIRES => {
      'Alien::libuv' => '1.000',
    },
    ...
  ),
);
```

# DESCRIPTION

This package can be used by other [CPAN](https://metacpan.org) modules that
require [libuv](http://libuv.org).

# BACKWARDS COMPATIBILITY

Each upstream release of [libuv](http://libuv.org) adds new features.
To ensure your `XS` or `FFI` code using this module can work with the various
inconsistencies from new features across the version 1 release, you might
consider a backwards compatibility header file akin to:

```
#include <string.h>
#include <stdint.h>
#include <uv.h>


/* pulled from sys/signal.h in case we don't have it in Windows */
#if !defined(SIGPROF)
#define SIGPROF 27 /* profiling time alarm */
#endif

/* added in 1.14 */
#if !defined(UV_DISCONNECT)
#define UV_DISCONNECT 4
#endif
#if !defined(UV_PRIORITIZED)
#define UV_PRIORITIZED 8
#endif
#if !defined(UV_VERSION_HEX)
#define UV_VERSION_HEX                                                         \
    ((UV_VERSION_MAJOR << 16) | (UV_VERSION_MINOR << 8) | (UV_VERSION_PATCH))
#endif

/* not added until v1.23 */
#if !defined(UV_PRIORITY_LOW)
#define UV_PRIORITY_LOW 19
#endif
#if !defined(UV_PRIORITY_BELOW_NORMAL)
#define UV_PRIORITY_BELOW_NORMAL 10
#endif
#if !defined(UV_PRIORITY_NORMAL)
#define UV_PRIORITY_NORMAL 0
#endif
#if !defined(UV_PRIORITY_ABOVE_NORMAL)
#define UV_PRIORITY_ABOVE_NORMAL -7
#endif
#if !defined(UV_PRIORITY_HIGH)
#define UV_PRIORITY_HIGH -14
#endif
#if !defined(UV_PRIORITY_HIGHEST)
#define UV_PRIORITY_HIGHEST -20
#endif

/* not added until 1.24 */
#if !defined(UV_PROCESS_WINDOWS_HIDE_CONSOLE)
#define UV_PROCESS_WINDOWS_HIDE_CONSOLE (1 << 5)
#endif
#if !defined(UV_PROCESS_WINDOWS_HIDE_GUI)
#define UV_PROCESS_WINDOWS_HIDE_GUI (1 << 6)
#endif

/* not available until 1.26 */
#if !defined(UV_THREAD_NO_FLAGS)
#define UV_THREAD_NO_FLAGS 0x00
#endif
#if !defined(UV_THREAD_HAS_STACK_SIZE)
#define UV_THREAD_HAS_STACK_SIZE 0x01
#endif
#if !defined(MAXHOSTNAMELEN)
#define MAXHOSTNAMELEN 255
#endif
#if !defined(UV_MAXHOSTNAMESIZE)
#define UV_MAXHOSTNAMESIZE (MAXHOSTNAMELEN + 1)
#endif
#if !defined(UV_IF_NAMESIZE)
#if defined(IF_NAMESIZE)
#define UV_IF_NAMESIZE (IF_NAMESIZE + 1)
#elif defined(IFNAMSIZ)
#define UV_IF_NAMESIZE (IFNAMSIZ + 1)
#else
#define UV_IF_NAMESIZE (16 + 1)
#endif
#endif
```

# AUTHOR

Chase Whitener <`capoeirab@cpan.org`>

# CONTRIBUTORS

Graham Ollis <`plicease@cpan.org`>

# COPYRIGHT & LICENSE

Copyright (c) 2017 Chase Whitener. All rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
Revision history for {{$dist->name}}

{{$NEXT}}
  - Add some documentation about backwards compatibility.
  - Update to the libuv v1.41.0 release.
  - Update to latest version of @Starter::Git for builds

1.014     2020-07-15
  - Update CI to use GitHub Actions rather than Travis and AppVeyor
  - Update the the libuv v1.38.1 release
  - Only run upstream (actual libuv tests) when RELEASE_TESTING
    or AUTOMATED_TESTING is on

1.013     2020-03-16
  - Update documentation and build to use Alien::Base::Wrapper
    This handles edge cases of prereqs that we didn't get manually
    prior to this update
  - Bump prereq to Alien::Build v2.15 to get Probe::Vcpkg
  - Move testing phase to the correct place for the library on non-Windows
  - Update to the libuv v1.34.2 release

1.012     2019-08-31
  - Update to the libuv v1.31.0 release
  - Run make check on non-windows environments
  - Update the dzil build
  - Stop CI testing on cygwin
  - Add CI tests for 5.30

1.011     2019-04-24
  - Update to the libuv v1.28.0 release

1.010     2019-04-10
  - Revert the removal of NetBSD as a fix for it is in the works:
    https://github.com/libuv/libuv/pull/2252

1.009     2019-04-09
  - Update to the libuv v1.27.0 release
  - Don't try to test on GNUkFreeBSD or NetBSD as the library isn't
    supported there

1.008     2019-02-07
  - Update to the libuv v1.25.0 release
  - Re-add the previously removed build flags for Windows

1.007     2019-01-13
  - Update to the libuv v1.24.1 release
  - Remove some unnecessary build flags
  - Add -msse2 flag

1.006     2018-08-28
  - Update to the libuv v1.23.0 release

1.005     2018-08-09
  - Don't even try to build on anything below Win7; It's broken.

1.004     2018-07-24
  - Update to the libuv v1.22.0 release
  - Deprecate all Windows versions prior to Windows 7 as per
    https://github.com/libuv/libuv/blob/v1.x/SUPPORTED_PLATFORMS.md
  - Switch to CMake when on Windows via Alien::cmake3
  - Add a missing -lshell32 library inclusion to the libs and static_libs
  - Add Perl 5.28 to the Strawberry tests on AppVeyor

1.003     2018-06-19
  - Update library inclusions for building with MinGW on Windows.
  - Add a contributing document.

1.002     2018-06-06
  - Update to the libuv v1.20.3 release

1.001     2018-04-15
  - Fix compiler flag typos for windows

1.000     2018-04-05
  - Update to the libuv v1.20.0 release
  - Bump to major version 1 to reflect same major version as libuv.
  - Update the alienfile to add the various libs needed on Win32
  - Update the alienfile to add the various cflags needed on Win32

0.017     2018-03-29
  - Update to the libuv v1.19.2 release

0.016     2018-01-29
  - Update to the libuv v1.19.1 release

0.015     2018-01-11
  - Update to the libuv v1.18.0 release

0.014     2017-11-27
  - Update to the libuv v1.17.0 release

0.013     2017-11-12
  - Update to the libuv v1.16.1 release

0.012     2017-10-03
  - Update to the libuv v1.15.0 release
  - Support for NetBSD

0.011     2017-09-06
  - Only require Alien::Autotools for a non-windows source build (Graham Ollis)
  - Update to the libuv v1.14.1 release.

0.010     2017-08-24
  - Production release with the changes from the three previous trial releases

0.009     2017-08-17 (TRIAL)
  - Re-release to fix the bsd-tar headers in the tarball

0.008     2017-08-17 (TRIAL)
  - Ensure the download URL for libuv is now over https as that's now default
  - Bump libuv version to 1.14.0

0.007     2017-08-15 (TRIAL)
  - Depend on Alien::Autotools now
  - Cleanup the build on Windows a bit (hopefully speed up this process)
  - Only require Alien::MSYS on Windows

0.006     2017-07-10
  - Bump the share version to libuv v1.13.1
  - Add a recommended test install of FFI::Platypus
  - Diag out the install_type for better debugging later

0.005     2017-07-08
  - Updated the requirements to include Alien::m4
  - Update requirements of Alien::libtool, automake, and autoconf for some
    corrections

0.004     2017-06-29
  - Ensure we're at least using libuv v1.0.0. Prior versions of libuv handled
    errors differently, causing problems for XS modules needing to deal with
    both.

0.003     2017-06-27
  - Bump prereqs for Alien:: tools
  - Isolate dynamic libraries (Graham Ollis)

0.002     2017-06-26
  - Use Alien::libtool (Graham Ollis)
  - Use Alien::autoconf (Graham Ollis)
  - Use Alien::automake (Graham Ollis)
  - Fix install on BSD, OS X, and Windows (Graham Ollis)

0.001     2017-06-16
  - initial version
  - Does not yet work on Windows.


Makefile.PL
# This Makefile.PL for Alien-libuv was generated by
# Dist::Zilla::Plugin::MakeMaker::Awesome 0.48.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

use ExtUtils::MakeMaker;

if ($^O eq 'MSWin32') {
    my $ver;
    my $err;
    { # catch block
        local $@;
        # try
        $err = $@ || 'Error' unless eval {
            require Win32;
            $ver = (Win32::GetOSVersion())[1];
            1;
        };
    }
    if ($err || $ver < 6) {
        die("OS unsupported\nWindows needs to be Win7 or higher.\n");
    }
}
elsif ($^O eq 'gnukfreebsd') {
    die("OS unsupported\n");
}

my %WriteMakefileArgs = (
  "ABSTRACT" => "Interface to the libuv library http://libuv.org",
  "AUTHOR" => "Chase Whitener <capoeirab\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Alien::Build" => "2.15",
    "Alien::Build::MM" => "0.32",
    "Alien::Build::Plugin::Build::Make" => 0,
    "Alien::Build::Plugin::Probe::Vcpkg" => "2.15",
    "Config" => 0,
    "ExtUtils::MakeMaker" => "6.52",
    "IPC::Cmd" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "Alien::Build" => "2.15",
    "Alien::Build::MM" => "0.32",
    "Alien::Build::Plugin::Build::Make" => "0.01",
    "Alien::Build::Plugin::Probe::Vcpkg" => "2.15",
    "ExtUtils::MakeMaker" => "6.52"
  },
  "DISTNAME" => "Alien-libuv",
  "LICENSE" => "perl",
  "NAME" => "Alien::libuv",
  "PREREQ_PM" => {
    "Alien::Base" => "2.15",
    "base" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.52",
    "File::Spec" => 0,
    "Test2::Suite" => 0,
    "Test2::V0" => 0,
    "Test::Alien" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "1.015",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

my %FallbackPrereqs = (
  "Alien::Base" => "2.15",
  "Alien::Build" => "2.15",
  "Alien::Build::MM" => "0.32",
  "Alien::Build::Plugin::Build::Make" => 0,
  "Alien::Build::Plugin::Probe::Vcpkg" => "2.15",
  "Config" => 0,
  "ExtUtils::MakeMaker" => "6.52",
  "File::Spec" => 0,
  "IPC::Cmd" => 0,
  "Test2::Suite" => 0,
  "Test2::V0" => 0,
  "Test::Alien" => 0,
  "Test::More" => "0.88",
  "base" => 0,
  "strict" => 0,
  "warnings" => 0
);

# BEGIN code inserted by Dist::Zilla::Plugin::AlienBuild
use Alien::Build::MM;
my $abmm = Alien::Build::MM->new;
%WriteMakefileArgs = $abmm->mm_args(%WriteMakefileArgs);
# END code inserted by Dist::Zilla::Plugin::AlienBuild

unless ( eval { ExtUtils::MakeMaker->VERSION('6.63_03') } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

if ( $^O !~ /win/i ) {
	$WriteMakefileArgs{PREREQ_PM}{'Alien::Autotools'} = $FallbackPrereqs{'Alien::Autotools'} = '1.00';
	$WriteMakefileArgs{PREREQ_PM}{'Alien::Build::Plugin::Build::Autoconf'} = $FallbackPrereqs{'Alien::Build::Plugin::Build::Autoconf'} = '0.04';
}

if ( $^O eq 'MSWin32' ) {
	$WriteMakefileArgs{PREREQ_PM}{'Alien::Build::Plugin::Build::CMake'} = $FallbackPrereqs{'Alien::Build::Plugin::Build::CMake'} = '0';
	$WriteMakefileArgs{PREREQ_PM}{'Alien::cmake3'} = $FallbackPrereqs{'Alien::cmake3'} = '0.04';
}

WriteMakefile(%WriteMakefileArgs);

# BEGIN code inserted by Dist::Zilla::Plugin::AlienBuild
{ package
    MY;
  sub postamble {
    $abmm->mm_postamble(@_);
  }
  sub install {
    $abmm->can('mm_install')
      ? $abmm->mm_install(@_)
      : shift->SUPER::install(@_);
  }
}
# END code inserted by Dist::Zilla::Plugin::AlienBuild


lib/Alien/libuv.pm
package Alien::libuv;

use strict;
use warnings;
use base qw( Alien::Base );

our $VERSION = '1.015';

1;

__END__

=head1 NAME

Alien::libuv - Interface to the libuv library L<http://libuv.org>

=head1 SYNOPSIS

In your C<Makefile.PL>:

    use strict;
    use warnings;

    use ExtUtils::MakeMaker;
    use Config;
    use Alien::Base::Wrapper ();

    WriteMakefile(
      Alien::Base::Wrapper->new('Alien::libuv')->mm_args2(
        ...
        CONFIGURE_REQUIRES => {
          'Alien::libuv' => '1.000',
        },
        ...
      ),
    );

=head1 DESCRIPTION

This package can be used by other L<CPAN|https://metacpan.org> modules that
require L<libuv|http://libuv.org>.

=head1 BACKWARDS COMPATIBILITY

Each upstream release of L<libuv|http://libuv.org> adds new features.
To ensure your C<XS> or C<FFI> code using this module can work with the various
inconsistencies from new features across the version 1 release, you might
consider a backwards compatibility header file akin to:

  #include <string.h>
  #include <stdint.h>
  #include <uv.h>


  /* pulled from sys/signal.h in case we don't have it in Windows */
  #if !defined(SIGPROF)
  #define SIGPROF 27 /* profiling time alarm */
  #endif

  /* added in 1.14 */
  #if !defined(UV_DISCONNECT)
  #define UV_DISCONNECT 4
  #endif
  #if !defined(UV_PRIORITIZED)
  #define UV_PRIORITIZED 8
  #endif
  #if !defined(UV_VERSION_HEX)
  #define UV_VERSION_HEX                                                         \
      ((UV_VERSION_MAJOR << 16) | (UV_VERSION_MINOR << 8) | (UV_VERSION_PATCH))
  #endif

  /* not added until v1.23 */
  #if !defined(UV_PRIORITY_LOW)
  #define UV_PRIORITY_LOW 19
  #endif
  #if !defined(UV_PRIORITY_BELOW_NORMAL)
  #define UV_PRIORITY_BELOW_NORMAL 10
  #endif
  #if !defined(UV_PRIORITY_NORMAL)
  #define UV_PRIORITY_NORMAL 0
  #endif
  #if !defined(UV_PRIORITY_ABOVE_NORMAL)
  #define UV_PRIORITY_ABOVE_NORMAL -7
  #endif
  #if !defined(UV_PRIORITY_HIGH)
  #define UV_PRIORITY_HIGH -14
  #endif
  #if !defined(UV_PRIORITY_HIGHEST)
  #define UV_PRIORITY_HIGHEST -20
  #endif

  /* not added until 1.24 */
  #if !defined(UV_PROCESS_WINDOWS_HIDE_CONSOLE)
  #define UV_PROCESS_WINDOWS_HIDE_CONSOLE (1 << 5)
  #endif
  #if !defined(UV_PROCESS_WINDOWS_HIDE_GUI)
  #define UV_PROCESS_WINDOWS_HIDE_GUI (1 << 6)
  #endif

  /* not available until 1.26 */
  #if !defined(UV_THREAD_NO_FLAGS)
  #define UV_THREAD_NO_FLAGS 0x00
  #endif
  #if !defined(UV_THREAD_HAS_STACK_SIZE)
  #define UV_THREAD_HAS_STACK_SIZE 0x01
  #endif
  #if !defined(MAXHOSTNAMELEN)
  #define MAXHOSTNAMELEN 255
  #endif
  #if !defined(UV_MAXHOSTNAMESIZE)
  #define UV_MAXHOSTNAMESIZE (MAXHOSTNAMELEN + 1)
  #endif
  #if !defined(UV_IF_NAMESIZE)
  #if defined(IF_NAMESIZE)
  #define UV_IF_NAMESIZE (IF_NAMESIZE + 1)
  #elif defined(IFNAMSIZ)
  #define UV_IF_NAMESIZE (IFNAMSIZ + 1)
  #else
  #define UV_IF_NAMESIZE (16 + 1)
  #endif
  #endif


=head1 AUTHOR

Chase Whitener <F<capoeirab@cpan.org>>

=head1 CONTRIBUTORS

Graham Ollis <F<plicease@cpan.org>>

=head1 COPYRIGHT & LICENSE

Copyright (c) 2017 Chase Whitener. All rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut
